File: dereplication_pilot/analyze_clusters.py
#!/usr/bin/env python3
"""Analyze an mmseqs easy-cluster _cluster.tsv output for cross-assembly redundancy.

Parses cluster membership, attributes each member back to its source assembly
group via its header prefix (headers in this project are already globally
unique and group-prefixed, e.g. mh_p110_000000000011, spaS30___0 -- no separate
provenance tagging needed), and reports how much of a participant's own
co-assembly + single-sample contigs/genes are redundant with each other.

See ../dereplication_brainstorm.md and /home/ljc444/.claude/plans/happy-moseying-clarke.md
for why this pilot exists.
"""
import argparse
import sys
from collections import defaultdict


def read_fasta_lengths(path):
    lengths = {}
    name = None
    length = 0
    with open(path) as fh:
        for line in fh:
            if line.startswith(">"):
                if name is not None:
                    lengths[name] = length
                name = line[1:].split()[0].strip()
                length = 0
            else:
                length += len(line.strip())
    if name is not None:
        lengths[name] = length
    return lengths


def group_of(seq_id, groups_sorted_desc):
    for g in groups_sorted_desc:
        if seq_id.startswith(g) and (len(seq_id) == len(g) or seq_id[len(g)] == "_"):
            return g
    return None


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--cluster-tsv", required=True)
    ap.add_argument("--fasta", required=True, help="pooled fasta used as clustering input (for lengths)")
    ap.add_argument("--groups", nargs="+", required=True)
    ap.add_argument("--tier", required=True, choices=["contigs", "orfs"])
    ap.add_argument("--participant", required=True)
    ap.add_argument("--out", required=True, help="cluster_membership.tsv output path")
    ap.add_argument("--summary", required=True, help="summary.txt output path")
    args = ap.parse_args()

    groups_sorted_desc = sorted(set(args.groups), key=len, reverse=True)
    lengths = read_fasta_lengths(args.fasta)

    clusters = defaultdict(list)  # rep_id -> [member_id, ...]
    unresolved = []
    with open(args.cluster_tsv) as fh:
        for line in fh:
            line = line.rstrip("\n")
            if not line:
                continue
            rep_id, member_id = line.split("\t")
            clusters[rep_id].append(member_id)

    rows = []
    per_group_total = defaultdict(int)
    per_group_total_bp = defaultdict(int)
    per_group_redundant = defaultdict(int)
    per_group_redundant_bp = defaultdict(int)

    n_singleton = 0
    n_multi = 0
    n_cross_group = 0
    total_seqs = 0
    total_bp = 0
    redundant_seqs = 0
    redundant_bp = 0

    for rep_id, members in clusters.items():
        member_groups = []
        for m in members:
            g = group_of(m, groups_sorted_desc)
            if g is None:
                unresolved.append(m)
                g = "UNKNOWN"
            member_groups.append(g)

        distinct_groups = set(member_groups)
        is_cross_group = len(distinct_groups) > 1

        if len(members) == 1:
            n_singleton += 1
        else:
            n_multi += 1
        if is_cross_group:
            n_cross_group += 1

        for m, g in zip(members, member_groups):
            length = lengths.get(m, 0)
            total_seqs += 1
            total_bp += length
            per_group_total[g] += 1
            per_group_total_bp[g] += length
            if is_cross_group:
                redundant_seqs += 1
                redundant_bp += length
                per_group_redundant[g] += 1
                per_group_redundant_bp[g] += length
            rows.append((rep_id, m, g, length, is_cross_group))

    with open(args.out, "w") as fh:
        fh.write("cluster_id\tmember_id\tsource_group\tlength\tcross_group_cluster\n")
        for rep_id, m, g, length, is_cross in rows:
            fh.write(f"{rep_id}\t{m}\t{g}\t{length}\t{int(is_cross)}\n")

    lines = []
    lines.append(f"=== {args.participant} / {args.tier} ===")
    lines.append(f"total sequences pooled: {total_seqs}  ({total_bp} bp/residues)")
    lines.append(f"total clusters: {len(clusters)}  (singleton: {n_singleton}, multi-member: {n_multi})")
    lines.append(f"cross-group clusters (redundant across >=2 assemblies): {n_cross_group}")
    if total_seqs:
        lines.append(
            f"redundant sequences: {redundant_seqs}/{total_seqs} "
            f"({100.0 * redundant_seqs / total_seqs:.1f}% by count, "
            f"{100.0 * redundant_bp / total_bp:.1f}% by bp)"
        )
    lines.append("per-group breakdown (redundant/total, count and bp):")
    for g in sorted(per_group_total):
        t, r = per_group_total[g], per_group_redundant[g]
        tb, rb = per_group_total_bp[g], per_group_redundant_bp[g]
        pct = 100.0 * r / t if t else 0.0
        pctb = 100.0 * rb / tb if tb else 0.0
        lines.append(f"  {g}: {r}/{t} ({pct:.1f}%) by count, {rb}/{tb} ({pctb:.1f}%) by bp")
    if unresolved:
        lines.append(f"WARNING: {len(unresolved)} member IDs did not match any known group prefix (first 5: {unresolved[:5]})")

    summary = "\n".join(lines) + "\n"
    with open(args.summary, "w") as fh:
        fh.write(summary)
    print(summary, end="")

    if unresolved:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: dereplication_pilot/R/analyze_clusters.R
#!/usr/bin/env Rscript
# R port of analyze_clusters.py -- analyzes an mmseqs easy-cluster _cluster.tsv output
# for cross-assembly redundancy. Parses cluster membership, attributes each member back
# to its source assembly group via its header prefix (headers in this project are
# already globally unique and group-prefixed), and reports how much of a participant's
# own co-assembly + single-sample contigs/genes are redundant with each other.
#
# Usage: Rscript R/analyze_clusters.R --cluster-tsv X --fasta Y --groups g1 g2 ...
#          --tier contigs|orfs --participant pNNN --out cluster_membership.tsv --summary summary.txt

suppressPackageStartupMessages({
  library(dplyr)
  library(readr)
  library(stringr)
})

# Minimal CLI parser (optparse isn't installed in this env): single-value flags take
# exactly the next token; --groups is the one multi-value flag (like argparse's
# nargs="+") and consumes every token up to the next --flag, matching how
# run_pilot.sh/run_sweep.sh invoke this with an unquoted variable expansion.
parse_args <- function(args) {
  single_flags <- c("cluster-tsv", "fasta", "tier", "participant", "out", "summary")
  out <- list()
  i <- 1
  while (i <= length(args)) {
    flag <- sub("^--", "", args[i])
    if (flag == "groups") {
      j <- i + 1
      vals <- character()
      while (j <= length(args) && !str_starts(args[j], "--")) {
        vals <- c(vals, args[j]); j <- j + 1
      }
      out$groups <- vals
      i <- j
    } else if (flag %in% single_flags) {
      out[[flag]] <- args[i + 1]
      i <- i + 2
    } else {
      i <- i + 1
    }
  }
  out
}

read_fasta_lengths <- function(path) {
  lines <- read_lines(path, progress = FALSE)
  header_idx <- which(str_starts(lines, ">"))
  names <- str_extract(lines[header_idx], "^>(\\S+)", group = 1)
  ends <- c(header_idx[-1] - 1, length(lines))
  seq_lines_per_header <- ends - header_idx  # number of sequence lines following each header
  # sum of nchar for the sequence lines belonging to each header, vectorized via a repeated group index
  group_id <- rep(seq_along(header_idx), times = seq_lines_per_header)
  seq_line_idx <- setdiff(seq_len(length(lines)), header_idx)
  seq_lens <- nchar(lines[seq_line_idx])
  lengths <- as.integer(tapply(seq_lens, group_id, sum))
  # tapply drops groups with zero sequence lines (empty record) -- restore as 0
  full <- integer(length(header_idx))
  full[as.integer(names(table(group_id)))] <- lengths
  setNames(full, names)
}

# Every real group name in this project matches ^(mh_p\d+|spaS\d+)$ -- a single
# vectorized regex extracts the owning group directly (same fix already applied in
# cross_participant_analysis.R, which found the naive loop-over-groups approach far too
# slow at millions of rows).
group_of <- function(seq_ids, groups_sorted_desc) {
  str_match(seq_ids, "^(mh_p\\d+|spaS\\d+)")[, 2]
}

opt <- parse_args(commandArgs(trailingOnly = TRUE))
opt$`cluster-tsv` -> cluster_tsv_path  # keep the hyphenated flag name readable below
groups <- unique(opt$groups)

lengths <- read_fasta_lengths(opt$fasta)

clusters <- read_tsv(cluster_tsv_path, col_names = c("rep_id", "member_id"), col_types = "cc", progress = FALSE) %>%
  mutate(source_group = group_of(member_id, groups),
         length = unname(lengths[member_id]))

unresolved <- sum(is.na(clusters$source_group))
clusters$source_group[is.na(clusters$source_group)] <- "UNKNOWN"

cluster_summary <- clusters %>%
  group_by(rep_id) %>%
  summarise(n_members = n(), n_distinct_groups = n_distinct(source_group), .groups = "drop") %>%
  mutate(is_cross_group = n_distinct_groups > 1)

clusters <- clusters %>% left_join(cluster_summary %>% select(rep_id, is_cross_group), by = "rep_id")

write_tsv(clusters %>% transmute(cluster_id = rep_id, member_id, source_group, length,
                                  cross_group_cluster = as.integer(is_cross_group)),
          opt$out)

total_seqs <- nrow(clusters)
total_bp <- sum(clusters$length)
n_singleton <- sum(cluster_summary$n_members == 1)
n_multi <- sum(cluster_summary$n_members > 1)
n_cross_group <- sum(cluster_summary$is_cross_group)
redundant <- clusters %>% filter(is_cross_group)
redundant_seqs <- nrow(redundant)
redundant_bp <- sum(redundant$length)

per_group <- clusters %>%
  group_by(source_group) %>%
  summarise(total_count = n(), total_bp = sum(length),
            redundant_count = sum(is_cross_group), redundant_bp = sum(length[is_cross_group]),
            .groups = "drop") %>%
  arrange(source_group)

lines <- c(
  sprintf("=== %s / %s ===", opt$participant, opt$tier),
  sprintf("total sequences pooled: %d  (%d bp/residues)", total_seqs, total_bp),
  sprintf("total clusters: %d  (singleton: %d, multi-member: %d)", nrow(cluster_summary), n_singleton, n_multi),
  sprintf("cross-group clusters (redundant across >=2 assemblies): %d", n_cross_group),
  if (total_seqs > 0) sprintf("redundant sequences: %d/%d (%.1f%% by count, %.1f%% by bp)",
                               redundant_seqs, total_seqs, 100 * redundant_seqs / total_seqs, 100 * redundant_bp / total_bp) else NULL,
  "per-group breakdown (redundant/total, count and bp):",
  sprintf("  %s: %d/%d (%.1f%%) by count, %d/%d (%.1f%%) by bp",
          per_group$source_group, per_group$redundant_count, per_group$total_count,
          100 * per_group$redundant_count / per_group$total_count,
          per_group$redundant_bp, per_group$total_bp,
          100 * per_group$redundant_bp / per_group$total_bp)
)
if (unresolved > 0) {
  lines <- c(lines, sprintf("WARNING: %d member IDs did not match any known group prefix", unresolved))
}

summary_text <- paste(lines, collapse = "\n")
writeLines(summary_text, opt$summary)
cat(summary_text, "\n")

if (unresolved > 0) quit(status = 1)
